pkg/devfile/parser/parse_test.go

```diff
@@ -2395,3 +2398,298 @@ func Test_parseParentAndPlugin_RecursivelyReference_withMultipleURI(t *testing.T
+				},
+			},
+		},
+	}
+
+	testServer := httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if strings.Contains(r.URL.Path, "notexist") {
+			w.WriteHeader(http.StatusNotFound)
+			return
+		}
+		var data []byte
+		var err error
+		if strings.Contains(r.URL.Path, "devfile.yaml") {
+			data, err = yaml.Marshal(relativeParentDevfile.Data)
+		} else {
+			data, err = yaml.Marshal(parentDevfile.Data)
+		}
+		if err != nil {
+			t.Errorf("unexpected error: %v", err)
+			return
+		}
+		_, err = w.Write(data)
+		if err != nil {
+			t.Errorf("unexpected error: %v", err)
+		}
+	}))
+	// create a listener with the desired port.
+	l, err := net.Listen("tcp", uri1)
+	if err != nil {
+		t.Errorf("unexpected error: %v", err)
+		return
+	}
+
+	// NewUnstartedServer creates a listener. Close that listener and replace
+	// with the one we created.
+	testServer.Listener.Close()
+	testServer.Listener = l
+
+	testServer.Start()
+	defer testServer.Close()
+
+	tests := []struct {
+		name        string
+		curDevfile  DevfileObj
+		uri         string
+		wantDevFile DevfileObj
+		wantErr     bool
+	}{
+		{
+			name: "case 1: should be able to parse from relative uri on local disk",
+			curDevfile: DevfileObj{
+				Ctx: devfileCtx.NewDevfileCtx(OutputDevfileYamlPath),
+				Data: &v2.DevfileV2{
+					Devfile: v1.Devfile{
+						DevWorkspaceTemplateSpec: v1.DevWorkspaceTemplateSpec{
+							Parent: &v1.Parent{
+								ImportReference: v1.ImportReference{
+									ImportReferenceUnion: v1.ImportReferenceUnion{
+										Uri: localRelativeURI,
+									},
+								},
+							},
+							DevWorkspaceTemplateSpecContent: v1.DevWorkspaceTemplateSpecContent{},
+						},
+					},
+				},
+			},
+			wantDevFile: localDevfile,
+			uri:         localRelativeURI,
+		},
+		{
+			name:        "case 2: should be able to parse relative uri from URL",
+			curDevfile:  parentDevfile,
+			wantDevFile: relativeParentDevfile,
+			uri:         localRelativeURI,
+		},
+		{
+			name: "case 3: should fail if no path or url has been set for devfile ctx",
+			curDevfile: DevfileObj{
+				Data: &v2.DevfileV2{
+					Devfile: v1.Devfile{
+						DevWorkspaceTemplateSpec: v1.DevWorkspaceTemplateSpec{
+							Parent: &v1.Parent{
+								ImportReference: v1.ImportReference{
+									ImportReferenceUnion: v1.ImportReferenceUnion{
+										Uri: localRelativeURI,
+									},
+								},
+							},
+							DevWorkspaceTemplateSpecContent: v1.DevWorkspaceTemplateSpecContent{},
+						},
+					},
+				},
+			},
+			wantErr: true,
+		},
+		{
+			name: "case 4: should fail if file not exist",
+			curDevfile: DevfileObj{
+				Ctx: devfileCtx.NewDevfileCtx(OutputDevfileYamlPath),
+				Data: &v2.DevfileV2{
+					Devfile: v1.Devfile{
+						DevWorkspaceTemplateSpec: v1.DevWorkspaceTemplateSpec{
+							Parent: &v1.Parent{
+								ImportReference: v1.ImportReference{
+									ImportReferenceUnion: v1.ImportReferenceUnion{
+										Uri: "notexist/devfile.yaml",
```

Comment:
`notExistURI`

Reply:
remove this one, as only ctx is required as you pointed out